### ISSUE TITLE
Should cpp-coveralls give up if there is no defined token ?

### ISSUE THREAD

**Issue by mkfifo** [eddyxu/cpp-coveralls#82]
Currently cpp-coveralls will still send data to coveralls even when no token is defined, which causes coveralls to reject the send (as it cannot associate it with a project).

I think cpp-coveralls should throw an exception just before it would send the report if there is no coveralls token defined, maybe this functionality could be hidden behind a flag.

This is because it took me a while to realise I needed to set a token, I would much prefer for it to throw an exception and explain how to resolve it.

I am happy to do this work if upstream will accept it.

**Comment by eddyxu** [comment]
@mkfifo Yeah, if you'd like, please feel free to do so.

**Comment by mkfifo** [comment]
@eddyxu are you happy for it to always give up if no token is found, or do you want it hidden behind a flag?

I would much prefer for it to always give up.

**Comment by eddyxu** [comment]
I'd prefer to raise Exception if no token was found. Most of the user should use `travis-ci`, so they are not affected. While for others, like you, it might be better to fix it other than hide it.

**Comment by mkfifo** [comment]
Awesome, I will get onto that shortly.

@eddyxu how are travis-ci users not affected by this?

**Comment by eddyxu** [comment]
@mkfifo  My mistake. Still in work now.  What I meant is that this token is required from `coveralls.io`, thus we should failure at the beginning when we find it is not set.

**Comment by mkfifo** [comment]
This has been added to pull request #81 which deals with issue #80 and this issue #82

**Comment by mkfifo** [comment]
for now I have rolled back the fix for issue #82 as it needs a little more work first

**Comment by mkfifo** [comment]
I have now re-pushed and updated PR #81 
I also had to change the testing file

**Comment by moteus** [comment]
For travis you does not have to set repo token.

**Comment by mkfifo** [comment]
@moteus how does that work?

**Comment by moteus** [comment]
[REFERENCING A REPOSITORY](https://coveralls.zendesk.com/hc/en-us/articles/201774865-API-Introduction)

**Comment by mkfifo** [comment]
@moteus thanks for that, I couldn't get that to work but maybe I was doing something incorrect.

In that case we do NOT want my code to raise an exception if no token is found, we only want that behavior in the case of running within a non-travis-ci environment.

This also changes my updates to the README file.

**Comment by mkfifo** [comment]
In that case I have removed the code that raises this error from my main master branch and pushed it into old-master

**Comment by mkfifo** [comment]
PR #81 now no longer includes anything touching this issue.

**Comment by mkfifo** [comment]
I would still like cpp-coveralls to issue an error IF there is no token, but as pointed out a token is only required in certain contexts and I think it is hard for cpp-coveralls to work out which context it is in.

I also still cannot seem to get travis-ci and coveralls.io behaving nicely together...

**Comment by mkfifo** [comment]
I would like to know how travis and cpp-coveralls work without any setup, I am using the free travis.

The instructions in coveralls.io tell me:

    First, if you are using Travis CI , make sure it's building correctly. If you're using Travis Pro, add the following lines to a .coveralls.yml file in your repository's root:
    service_name: travis-pro
    repo_token: TOKEN_GOES_HERE

I tried having a travis.yml matching the one given in [1] with no other travis setup mentioning coveralls but I did not see the build updates make it to coveralls.

For example my project linear_hash has a travis.yml matching the example [2], my travis page is [3] and my coveralls page is [4]

The commit currently missing on coveralls.io (but that is present on github and travis) is [5]

I can see that the travis build has completed, but the coveralls page has NOT yet updated.

Am I missing some setup in travis.ci for automated coveralls.io population?
Do I have to wait longer?, as generally my coverall.io jobs appear immediately.

[1] https://github.com/eddyxu/cpp-coveralls#linux
[2] https://github.com/mkfifo/linear_hash/blob/master/.travis.yml
[3] https://travis-ci.org/mkfifo/linear_hash/builds/59188736
[4] https://coveralls.io/r/mkfifo/linear_hash?branch=master
[5] https://github.com/mkfifo/linear_hash/commit/fc534596d2bebec149a706d860e57da9d318eeaa

**Comment by eddyxu** [comment]
@mkfifo It seems that it requires either `service_job_id`, which is `TRAVIS_JOB_ID`, or a `repo token`. 

>We can reference a repository in one of two ways:
> 1. service_job_id String and service_name String
> 2. repo_token String

Sorry, I have not closely followed `coveralls.io` for a while...

Are you working on this right now? Otherwise, I can fix this shortly

**Comment by mkfifo** [comment]
@eddyxu thanks but that doesn't seem to be working for me, assuming the job id is automatically setup.

If you compare [1] to [2] that job is completely missing from coverall.io

So I am wondering what I need to do to have this automatically work?

As for the pull request what this means is I can have my code throw an exception if both the service_jobs_id AND the repo_token are not defined.

[1] https://travis-ci.org/mkfifo/linear_hash/builds/59188736
[2] https://coveralls.io/r/mkfifo/linear_hash?branch=master

**Comment by eddyxu** [comment]
There is actually a result for your travis job  on coveralls.io

https://coveralls.io/jobs/5536318

Could you try to run it with `--verbose`

**Comment by mkfifo** [comment]
@eddyxu thanks for that, I am not seeing it on [1]

Running cpp-coverals with `--verbose` I get [2]

[1] https://coveralls.io/r/mkfifo/linear_hash?branch=master
[2] https://travis-ci.org/mkfifo/linear_hash/builds/59193237

**Comment by eddyxu** [comment]
@mkfifo  Take a look of https://coveralls.io/r/mkfifo/linear_hash?page=3  

Build #12 11 and such are actually the results of the recent build. I am not sure whether it is a coveralls.io issue...

**Comment by mkfifo** [comment]
@eddyxu thanks for that, if you look at [1] you can see the job number is "Job #11.1" and for [2] it is "Job #12.1" so whatever is generating this field is wrong:

    {u'url': u'https://coveralls.io/jobs/5536318', u'message': u'Job #11.1'}

[1] https://travis-ci.org/mkfifo/linear_hash/builds/59188736#L328
[2] https://travis-ci.org/mkfifo/linear_hash/builds/59193237#L330

**Comment by mkfifo** [comment]
Which I think is the result of the final `print(result)` in report.py `post_report`

```
def post_report(coverage):
    """Post coverage report to coveralls.io."""
    response = requests.post(URL, files={'json_file': json.dumps(coverage)})
    try:
        result = response.json()
    except ValueError:
        result = {'error': 'Failure to submit data. '
                  'Response [%(status)s]: %(text)s' % {
                      'status': response.status_code,
                      'text': response.text}}
    print(result)
    if 'error' in result:
        return result['error']
    return 0
```

Which makes me suspect a coveralls.io bug

**Comment by mkfifo** [comment]
@moteus @eddyxu I have address the concerns around breaking the existing travis workflow and made a PR #84 which will raise an exception only if it cannot find the repo token and if it also cannot find the travis job id.

My issue reported above is still outstanding but I suspect that is a coverall.io bug.

**Comment by mkfifo** [comment]
This issue was resolved in PR #84 

The issue I was having is a bug with coveralls.io, unsure how to report it, but it doesn't change cpp-coveralls.
